src/index.tsx

```diff
@@ -78,15 +78,37 @@ export const useIntersectionObserver = ({
 }: UseIntersectionObserverProperties) => {
   const [inView, setInView] = useState(false)
 
+  // We need to track if the callback has been triggered since the Observer
+  // callback will always be immediately invoked on page load.
+  // We don't want the immediate invocation to prevent the initial intersection
+  // callback, so we need to manually track it ourselves.
+  const hasRunCallbackOnceRef = useRef(false)
+
   const handleIntersect = (entries: IntersectionObserverEntry[]) => {
     if (!intersectObs) return
 
-    const [entry] = entries
+    // Capture the mutable ref for this closure.
+    const hasRunCallbackOnce = hasRunCallbackOnceRef.current
+
+    // We've already ran the callback and triggerOnce is true, so don't
+    // do anything.
+    if (hasRunCallbackOnce && options.triggerOnce) return
+
+    // Otherwise, we need to see if the element is intersecting and run
+    // the user's callback if they have provided one.
+    const someElementIsInView = entries.some(e => e.isIntersecting)
 
-    if (callback && entry.isIntersecting) callback(entry)
-    if (options.triggerOnce && entry.isIntersecting) intersectObs.disconnect()
+    if (someElementIsInView) {
+      if (callback) callback(entries)
+      hasRunCallbackOnceRef.current = true
+    }
 
-    setInView(entry.isIntersecting)
+    // If triggerOnce is true and we've already ran the callback,
+    // disconnect so we don't trigger anymore.
+    if (options.triggerOnce && hasRunCallbackOnceRef.current)
+      intersectObs.disconnect()
+
+    setInView(someElementIsInView)
```

Comment:
I would not do this. In case of `triggerOnce = false`, this might produce wrong results in case the `entries` array contains mixed `isIntersecting` entries, and the last entry in the array is `isIntersecting = false`.

Reply:
That‘s why I would treat the `triggerOnce` case differently alltogether as described in https://github.com/asyarb/use-intersection-observer/issues/5#issue-580645869

Reply:
Gotcha. After being able to reliably reproduce the issue I'm totally seeing what you're getting at. (needed to add over 100 observed keys and scroll **really** fast 😅)

Changes can be seen in #5697e7f.